### PR TITLE
Add new registration email for new users.

### PR DESCRIPTION
**Pull request by LinusFresz**
As a follow-up to https://github.com/thewca/worldcubeassociation.org/issues/1814 I've looked into creating a new confirmation mail for newly created user accounts.

However, we can only change the reconfirmation instructions mail (which was originally requested in https://github.com/thewca/worldcubeassociation.org/issues/1814 if we overwrite `send_reconfirmation_instructions` in our User model as well. To be honest, I am not entirely sure if it's worth it to only change a few words.

I've already added tests (which are going to fail for the foreign language mails) to test the new registration mail. I used https://github.com/thewca/worldcubeassociation.org/blob/f66b258c7d8f54defc0544d912c6d521f99fc316/WcaOnRails/spec/mailers/registrations_mailer_spec.rb Lines 26 - 46 as comparison to create a test for foreign languages as well, but the test for competition registration mails obviously assumes that we have translations for the confirmation mails. How can I avoid to run into test errors for these language tests while we only have a user registration confirmation mail in english?

I am definitely open for discussions about the mail text! :)

Screenshot:
new-registration-mail

### CODE REVIEW

File: WcaOnRails/config/locales/en.yml [hunk 1]
```diff
@@ -602,6 +602,15 @@ en:
...
       unavailable: "Requested locale is not available."
+    #context: user creating new account and receiving confirmation mail
+    mailer:
+      create_new_account:
+        header: "Confirmation instructions"
+        welcome: "Welcome %{user_email}!"
+        confirm: "Your account is created and must be activated before you can use it. Please confirm your account through the link below: "
+        activate: "After activation you may login to the website of the World Cube Association: "
+        addendum: "This email is generated automatically. If you didn't register an account on the website of the World Cube Association and received this message by mistake, please ignore and delete it."
+        questions: "Do you have more questions? We will be happy to answer them! Feel free to send a mail to "
```
Comment (jfly) [inline WcaOnRails/config/locales/en.yml]: Translations like this are always concerning to me, because it assumes that the language is going to want this link or whatever to come at the end of the sentence. We prefer to use string interpolation whenever possible, so this would become something like:

```
questions: "Do you have more questions? We will be happy to answer them! Feel free to send a mail to %{contact_link}."
```

After this change, you would also have to change view code a bit.

I think you could make this same sort of change for the `activate` rule above.

File: WcaOnRails/spec/mailers/new_registration_mailer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      expect(mail.from).to eq ["notifications@worldcubeassociation.org"]
+    end
+
+    it "renders the body" do
+      expect(mail.body.encoded).to match user.email
+      expect(mail.body.encoded).to match "Your account is created and must be activated before you can use it."
+    end
+  end
+
+  describe "send registration mail to new users in their language" do
```
Comment (jfly) [inline WcaOnRails/spec/mailers/new_registration_mailer_spec.rb]: I don't think it's terribly important to test this. As you already mentioned, these tests won't actually pass until the French translations are done anyways.

This sort of test is from the early days of internationalization on the WCA website when we were still trying to figure out how to do it right (that `localized_mail` method you're using is what we ended up building to make this easier on us).

File: WcaOnRails/app/models/user.rb [hunk 3]
```diff
@@ -331,6 +331,10 @@ def avatar_requires_wca_id
     end
   end
 
+  def send_on_create_confirmation_instructions
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: I think it would be good to have a test that confirms that this method actually gets called when a user registers. I think that would look like a test in `user_spec` that mocks out `NewRegistrationMailer.send_registration_mail` and asserts that it is called with this new user.

File: WcaOnRails/spec/mailers/new_registration_mailer_spec.rb (hunk 2 above)
Comment (LinusFresz) [reply WcaOnRails/spec/mailers/new_registration_mailer_spec.rb]: Alright, thanks for letting me know! I just hoped there is some way to check if either a translation is available or use the English text instead. But as it's not, I'll remove the translation test.

File: WcaOnRails/app/models/user.rb (hunk 3 above)
Comment (LinusFresz) [reply WcaOnRails/app/models/user.rb]: Fixed tests and mail text!

I also created all necessary files for the new email address change confirmation mail as requested in https://github.com/thewca/worldcubeassociation.org/issues/1814. 

Screenshot:
new-mail-confirmation

**Comment by LinusFresz** [comment]
sorry for the still failing tests, I am going to fix them tomorrow.

**Comment by LinusFresz** [comment]
Ok, so tests are fixed now and travis build passed! However it says "Coverage decreased", what exactly does it mean? Do I have to take any action?

File: WcaOnRails/spec/models/user_spec.rb [hunk 4]
```diff
@@ -353,6 +353,27 @@
     end
   end
 
+  describe "check if registration form sends mail to newly registered user" do
+    it "send mail" do
+      new_user = FactoryBot.create :user, :unconfirmed
+      expect(ActionMailer::Base.deliveries.last.to).to eq [new_user.email]
```
Comment (jfly) [inline WcaOnRails/spec/models/user_spec.rb]: This is an interesting way of checking if the email was sent! However, we haven't been checking for emails this way, as it's a little brittle. You already tested that the delivery address and subject are correct in your mailer spec. Instead, we've been using mocks. If you look in this file, you'll see https://github.com/jfly/worldcubeassociation.org/blob/ebb8e956eef9ea57b112333f8ba5835ba0193731/WcaOnRails/spec/models/user_spec.rb#L498, where we check that the right mailer method was called with the correct arguments.

File: WcaOnRails/spec/models/user_spec.rb [hunk 5]
```diff
@@ -353,6 +353,27 @@
...
+
+  describe "let user change his mail address" do
+    it "send mail" do
+      user = FactoryBot.create :user
+      user.unconfirmed_email = "new@example.com"
+      expect(user.pending_reconfirmation?).to eq true
+      user.send_reconfirmation_instructions
+
+      expect { user.send_reconfirmation_instructions } .to change { ActionMailer::Base.deliveries.count }.by(1)
+      expect(ActionMailer::Base.deliveries.last.to).to eq [user.unconfirmed_email]
```
Comment (jfly) [inline WcaOnRails/spec/models/user_spec.rb]: As I commented above, I think it would be better to check a mock here, rather than looking at the actual email that was sent.

File: WcaOnRails/spec/models/user_spec.rb [hunk 6]
```diff
@@ -353,6 +353,27 @@
...
+      expect(ActionMailer::Base.deliveries.last.subject).to eq "Confirmation instructions"
+    end
+  end
+
+  describe "let user change his mail address" do
+    it "send mail" do
+      user = FactoryBot.create :user
+      user.unconfirmed_email = "new@example.com"
+      expect(user.pending_reconfirmation?).to eq true
+      user.send_reconfirmation_instructions
```
Comment (jfly) [inline WcaOnRails/spec/models/user_spec.rb]: Is this really supposed to be here? I'm surprised because you call the same method on the next line inside of the `expect` block.

File: WcaOnRails/spec/models/user_spec.rb [hunk 7]
```diff
@@ -353,6 +353,27 @@
...
+
+      expect { user.send_reconfirmation_instructions } .to change { ActionMailer::Base.deliveries.count }.by(1)
```
Comment (jfly) [inline WcaOnRails/spec/models/user_spec.rb]: Instead of explicitly calling `user.send_reconfirmation_instructions`, can you instead do a `user.save!` here? The problem with the test as currently written is that it checks that the `send_reconfirmation_instructions` method does the right thing, but it doesn't check that devise actually calles the `send_reconfirmation_instructions` method.

File: WcaOnRails/app/views/new_registration_mailer/send_registration_mail.html.erb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+</p>
+
+<p>
+  <%= link_to t('users.mailer.create_new_account.confirmation_link'), confirmation_url(@user, confirmation_token: @user.confirmation_token, locale: I18n.locale) %>
+</p>
+
+<p>
+  <%= t 'users.mailer.create_new_account.activate' %>
+
+  <%= link_to('login form', 'https://www.worldcubeassociation.org/users/sign_in') %>
```
Comment (jfly) [inline WcaOnRails/app/views/new_registration_mailer/send_registration_mail.html.erb]: nit: This won't work for environments with a different url (such as staging or local development). It would be better to use `new_user_registration_url` here instead.

File: WcaOnRails/config/locales/en.yml [hunk 9]
```diff
@@ -602,6 +602,23 @@ en:
...
       unavailable: "Requested locale is not available."
+    #context: user receiving mails
+    mailer:
+      #context: user creating new account and receiving confirmation mail
+      create_new_account:
+        header: "Confirmation instructions"
+        welcome: "Welcome %{user_email}!"
+        confirm: "Your account is created and must be activated before you can use it. Please confirm your account through the link below: "
+        confirmation_link: "Confirm my account."
+        activate: "After activation you may login to the website of the World Cube Association: "
```
Comment (jfly) [inline WcaOnRails/config/locales/en.yml]: nit: login -> log in. "login" is a noun, "log in" is the verb phrase.

**Comment by jfly** [comment]
> However it says "Coverage decreased", what exactly does it mean? Do I have to take any action?

We use a tool to keep track of how much of our real code is actually executed when we run our tests. The idea is that if we write a whole bunch of code, and none of our tests actually execute that code, that's a sign that new code is probably untested and could easily break as future people work on it.

I think your test coverage here is pretty good. I'm not sure why exactly the coverage when down by 0.01%, but I think it's ok to ignore here.

File: WcaOnRails/app/models/user.rb [hunk 10]
```diff
@@ -331,6 +331,18 @@ def avatar_requires_wca_id
...
+    NewRegistrationMailer.send_registration_mail(self).deliver_now
+  end
+
+  def send_reconfirmation_instructions
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: Can you add a comment here about how this was copied from https://github.com/plataformatec/devise/blob/master/lib/devise/models/confirmable.rb#L124?

File: WcaOnRails/app/models/user.rb [hunk 11]
```diff
@@ -331,6 +331,18 @@ def avatar_requires_wca_id
...
+    @reconfirmation_required = false
+
+    unless @skip_confirmation_notification
+      ReconfirmationMailer.send_reconfirmation_mail(self).deliver_now
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: Actually, I'm concerned that this is going to break reconfirmation. This line of code looks a lot different than the `def send_confirmation_instructions` https://github.com/plataformatec/devise/blob/b8f644198d34e96039b84b69842ecb4edffa4664/lib/devise/models/confirmable.rb#L115-L122 that is currently happening.

Can we simply not override this method, and instead edit WcaOnRails/app/views/devise/mailer/confirmation_instructions.html.erb?

File: WcaOnRails/spec/models/user_spec.rb (hunk 6 above)
Comment (LinusFresz) [reply WcaOnRails/spec/models/user_spec.rb]: Woops, missed this one from my previous work, will simply remove this line.

File: WcaOnRails/app/models/user.rb (hunk 10 above)
Comment (LinusFresz) [reply WcaOnRails/app/models/user.rb]: Sure, will do!

File: WcaOnRails/app/models/user.rb (hunk 11 above)
Comment (LinusFresz) [reply WcaOnRails/app/models/user.rb]: the lines of code you are talking about are actually `send_*re*confirmation_instructions`, not `send_confirmation_instructions`. And besides the changed method call, the methods are identical! I did not change anything about `send_confirmation_instructions`, so we're still using Devise's method for this, but `send_reconfirmation_instructions` does not call `send_confirmation_instructions` anymore but the newly created mailer instead.

Comment (jfly) [reply WcaOnRails/app/models/user.rb]: Sorry, my comment wasn't very clear. Let me try again:

The method you've written here looks identical to devise's, **except** for the fact that devises's calls `send_confirmation_instructions`, and yours calls `ReconfirmationMailer.send_reconfirmation_mail(self).deliver_now`. I'm concerned that this is not a safe change, because `send_confirmation_instructions` looks like it does a number of things (such as sometimes generating a confirmation token?), and the new code you've written doesn't do that.

It doesn't feel like this `send_reconfirmation_instructions` method is supposed to be overridden.

**Comment by LinusFresz** [comment]
I've added some fixes! Most importantly I removed `send_reconfirmation_instructions` and used the 'confirmation mail' template from Devise for the email reconfirmation.

Note, that I've changed the `:unconfirmed` trait into a transient `confirmed` in users.rb to avoid automatic confirmation if needed (as for my case). Only one other test seemed to used this trait and should be fixed as well.

File: WcaOnRails/app/models/user.rb [hunk 12]
```diff
@@ -331,6 +331,11 @@ def avatar_requires_wca_id
     end
   end
 
+  # The methods send_on_create_confirmation_instructions and send_reconfirmation_instructions were copied from https://github.com/plataformatec/devise/blob/master/lib/devise/models/confirmable.rb
```
[the remark this answers is not in the document]
Comment (LinusFresz) [reply WcaOnRails/app/models/user.rb]: well, technically `send_on_create_confirmation instructions` is still overridden from Devise, so I am going to edit it.

**Comment by jfly** [comment]
Awesome work here, @LinusFresz!
